[quantization] Store bias in PackedConvWeight in fbgemm

supriyar opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#25626 [quantization] Store bias in PackedConvWeight in fbgemm**

Add bias as an optional parameter in the packed conv weight struct.

Differential Revision: [D17177723](https://our.internmc.facebook.com/intern/diff/D17177723/)

supriyar committed:
[quantization] Store bias in PackedConvWeight in fbgemm

Add bias as an optional parameter in the packed conv weight struct.

Differential Revision: [D17177723](https://our.internmc.facebook.com/intern/diff/D17177723/)

supriyar committed:
Update on "[quantization] Store bias in PackedConvWeight in fbgemm"


Add bias as an optional parameter in the packed conv weight struct.

Differential Revision: [D17177723](https://our.internmc.facebook.com/intern/diff/D17177723/)

supriyar committed:
Update on "[quantization] Store bias in PackedConvWeight in fbgemm"


Add bias as an optional parameter in the packed conv weight struct.

Differential Revision: [D17177723](https://our.internmc.facebook.com/intern/diff/D17177723/)

dskhudia commented on torch/nn/_intrinsic/quantized/modules/conv_relu.py:
```diff
@@ -24,29 +24,13 @@ def __init__(self, in_channels, out_channels, kernel_size, stride=1,
                                          stride=stride, padding=padding, dilation=dilation,
                                          groups=groups, bias=bias, padding_mode=padding_mode)
 
-    def weight(self):
```
Are these two methods being removed?

supriyar replied:
Yes, I don't see why they are needed as the parent module already defined them. Looks like linear_relu.py also doesn't have them

supriyar commented on aten/src/ATen/native/quantized/cpu/qconv.cpp:
```diff
@@ -124,20 +123,38 @@ class QConv2dInt8 final : public c10::OperatorKernel {
+        for (int i = 0; i < K; ++i) {
+          bias_scale.data_ptr<double>()[i] = pack_ptr.w_scale[i] * act_scale;
+        }
+        qbias = quantize_linear_per_channel_cpu(
```
Using the `quantize_linear_per_channel` function causes the following error 
```
RuntimeError: Expected object of type Variable but found type CPUFloatType for argument #1 'scales' (checked_cast_variable at caffe2/torch/csrc/autograd/VariableTypeManual.cpp:38)
```
Spoke to @yf225 offline and he said the error suggests that we are missing a at::AutoNonVariableTypeMode thread-local guard somewhere in the call path of QLinearInt8 operator().
quantize_linear_per_channel dispatches to VariableType dispatcher, which expects the input tensor to be a Variable.

jerryzh168 replied:
do you know why are we missing this guard?

supriyar replied:
I am not sure. @smessmer any thoughts on if this guard should be added?

raghuramank100 replied:
Is this issue resolved?

supriyar replied:
Not yet, I will file a separate issue for this.

supriyar commented on aten/src/ATen/native/quantized/cpu/qconv.cpp:
```diff
@@ -124,20 +123,38 @@ class QConv2dInt8 final : public c10::OperatorKernel {
+        qbias = at::quantize_linear(
+            at::dequantize(bias), pack_ptr.w_scale[0] * act_scale, 0, kQInt32);
+      } else if (pack_ptr.q_scheme == kPerChannelAffine) {
+        std::vector<int64_t> vect(1, 0);
```
Done.

supriyar committed:
Update on "[quantization] Store bias in PackedConvWeight in fbgemm"


Add bias as an optional parameter in the packed conv weight struct.

Differential Revision: [D17177723](https://our.internmc.facebook.com/intern/diff/D17177723/)

supriyar committed:
Update on "[quantization] Store bias in PackedConvWeight in fbgemm"


Add bias as an optional parameter in the packed conv weight struct.

Differential Revision: [D17177723](https://our.internmc.facebook.com/intern/diff/D17177723/)

supriyar committed:
Update on "[quantization] Store bias in PackedConvWeight in fbgemm"


Add bias as an optional parameter in the packed conv weight struct.

Differential Revision: [D17177723](https://our.internmc.facebook.com/intern/diff/D17177723/)

supriyar committed:
Update on "[quantization] Store bias in PackedConvWeight in fbgemm"


Add bias as an optional parameter in the packed conv weight struct.

Differential Revision: [D17177723](https://our.internmc.facebook.com/intern/diff/D17177723/)

supriyar committed:
Update on "[quantization] Store bias in PackedConvWeight in fbgemm"


Add bias as an optional parameter in the packed conv weight struct.

Differential Revision: [D17177723](https://our.internmc.facebook.com/intern/diff/D17177723/)

supriyar committed:
Update on "[quantization] Store bias in PackedConvWeight in fbgemm"


Add bias as an optional parameter in the packed conv weight struct.

Differential Revision: [D17177723](https://our.internmc.facebook.com/intern/diff/D17177723/)

raghuramank100 commented on test/test_quantized.py:
```diff
@@ -1136,9 +1135,9 @@ def test_qconv_unpack(self, X, strideH, strideW, padH, padW, channelwise):
+        W_packed = qconv_prepack(W_q, None, strides, paddings, dilations, groups)
         # Unpack weights weight unpacking operator (Used for serialization)
-        W_unpacked = qconv_unpack(W_packed)
+        W_unpacked = qconv_unpack(W_packed)[0]
```
Please add a check to ensure that the unpacked bias is identical to the original bias.

raghuramank100 reviewed (approved):
Looks great, please address comments prior to submitting.

supriyar committed:
Update on "[quantization] Store bias in PackedConvWeight in fbgemm"


Add bias as an optional parameter in the packed conv weight struct.

Differential Revision: [D17177723](https://our.internmc.facebook.com/intern/diff/D17177723/)
